File: dotnet/MoviesSparkDemo.cs
namespace Southworks.Demos.Spark
{
    using System;
    using System.Collections.Generic;
    using System.Diagnostics;
    using System.Linq;
    using Microsoft.Spark.Sql;
    using Microsoft.Spark.Sql.Types;
    using Southworks.Demos.Spark.Metrics;

    public static class MoviesSparkDemo
    {
        private const string MetricsToObtain = "metricsToObtain";
        private const string LogLevel = "logLevel";
        private const string MoviesCsvFile = "dataset/movies.csv";
        private const string RatingsCsvFile = "dataset/ratings.csv";
        private const string SparkAppName = "MoviesSparkDemo";

        public static void Main(string[] args)
        {
            var parameters = ParseArguments(args);
            var logLevel = (string) parameters[LogLevel];
            var metricsToObtain = (List<string>) parameters[MetricsToObtain];
            Run(logLevel, metricsToObtain);
        }

        private static Dictionary<string, object> ParseArguments(IEnumerable<string> args)
        {
            var argsMap = new Dictionary<string, object>
            {
                // Default parameters: logLevel=WARN, metrics = all.
                {LogLevel, "WARN"}, {MetricsToObtain, new List<string>(new[] {"1", "2", "3", "4", "5", "6"})}
            };

            // This iteration allows us to make parameters optional and position-agnostic.
            foreach (var parameter in args)
            {
                if (parameter.ToUpper().Contains("LOGLEVEL="))
                {
                    argsMap[LogLevel] = parameter.Split('=')[1].ToUpper();
                }

                if (parameter.ToUpper().Contains("METRICS="))
                {
                    argsMap[MetricsToObtain] = new List<string>(parameter.Split('=')[1].Split(','));
                }
            }

            return argsMap;
        }

        private static void Run(string logLevel, List<string> metricsToObtain)
        {
            var spark = SparkSession
                .Builder()
                .AppName(SparkAppName)
                .GetOrCreate();

            spark.SparkContext.SetLogLevel(logLevel);

            // Read initial dataframes as non-streaming DataFrames
            var moviesDataFrame = ReadCsvIntoDataframe(spark, MoviesCsvFile, SchemaLoader.MovieSchema);

            // Read initial dataframes as non-streaming DataFrames
            var ratingsDataFrame = ReadCsvIntoDataframe(spark, RatingsCsvFile, SchemaLoader.RatingSchema);

            foreach (var metric in metricsToObtain)
            {
                var watch = new Stopwatch();
                watch.Start();
                var colRows = RunMetric(moviesDataFrame, ratingsDataFrame, metric);
                watch.Stop();

                PrintRows(colRows.ToList(), watch.Elapsed.TotalSeconds);
            }
        }

        private static IEnumerable<Row> RunMetric(DataFrame moviesDataFrame, DataFrame ratingsDataFrame, string metric)
        {
            switch (metric)
            {
                case "1":
                    Console.WriteLine("\nBest films by overall rating");
                    return new BestFilmsByOverallRating().Run(moviesDataFrame, ratingsDataFrame).ToLocalIterator();
                case "2":
                    Console.WriteLine("\nMost rated films");
                    return new MostRatedFilms().Run(moviesDataFrame, ratingsDataFrame).ToLocalIterator();
                case "3":
                    Console.WriteLine("\nGlobal ratings given by each user per category");
                    return new GlobalRatingsGivenByEachUserPerCategory().Run(moviesDataFrame, ratingsDataFrame)
                        .ToLocalIterator();
                case "4":
                    Console.WriteLine("\nUsers giving the lowest ratings");
                    return new UsersGivingTheLowestRatings().Run(moviesDataFrame, ratingsDataFrame).ToLocalIterator();
                case "5":
                    Console.WriteLine("\nGenres by average rating");
                    return new GenresByAverageRating().Run(moviesDataFrame, ratingsDataFrame).ToLocalIterator();
                case "6":
                    Console.WriteLine("\nGenres combinations");
                    return new GenresCombinations().Run(moviesDataFrame, ratingsDataFrame).ToLocalIterator();
                default:
                    var sExceptionMessage = "Tried to obtain an unavailable metric: " + metric.Trim();
                    Console.WriteLine(sExceptionMessage);
                    throw new Exception(sExceptionMessage);
            }
        }

        private static DataFrame ReadCsvIntoDataframe(SparkSession sparkSession, string filename, StructType schema)
        {
            return sparkSession.Read()
                .Format("csv")
                .Option("header", "true").Schema(schema)
                .Load(filename);
        }

        private static void PrintRows(IReadOnlyList<Row> results, double elapsed)
        {
            var elapsedTime = string.Format("{0:0.00}", elapsed);
            Console.WriteLine("Results obtained in " + elapsedTime + "s\n");

            var aFieldNames = results[0].Schema.Fields;
            var headers = new List<string>();
            foreach (var s in aFieldNames)
            {
                headers.Add(s.Name);
            }

            Console.WriteLine(string.Join("\t - \t", headers));

            foreach (var row in results)
            {
                Console.WriteLine(string.Join("\t - \t", row.Values));
            }
        }
    }
}

File: python/MoviesSparkDemo.py
import sys
import os
import time
from pyspark.sql import SparkSession
import Metrics
import SchemaLoader

MOVIES_CSV_FILE = "../dataset/movies.csv"
RATINGS_CSV_FILE = "../dataset/ratings.csv"
SPARK_APP_NAME = "MoviesSparkDemo"

# Application entry point. Responsible for initialization issues, like parameters handling.
# No more behavior should be placed here.
# @param args Application parameters. Accepted: [logLevel=< INFO|ERROR|WARN|FINE >] [metrics=< 1|2|3|4|5|6 >[,1|2|3|4|5|6]]
def main():
    parameters = obtainApplicationArguments(sys.argv)
    sLogLevel = parameters["logLevel"]
    metricsToObtain = parameters["metricsToObtein"]
    run(sLogLevel, metricsToObtain)

# Run and display the specified metrics.
def run(sLogLevel, metrics):
    #Build spark session
    spark = SparkSession \
        .builder \
        .appName(SPARK_APP_NAME) \
        .getOrCreate()

    spark._sc.setLogLevel(sLogLevel)

    #Read initial datasets as non-streaming DataFrames
    moviesDataset = readCsvIntoDataframe(
        spark, MOVIES_CSV_FILE, SchemaLoader.getMovieSchema())
    ratingsDataset = readCsvIntoDataframe(
        spark, RATINGS_CSV_FILE, SchemaLoader.getRatingSchema())

    for metric in metrics:
        dStart = time.time()
        results = runMetric(moviesDataset, ratingsDataset, metric)
        dEnd = time.time()

        printResults(results, dStart, dEnd)

def obtainApplicationArguments(parameters):
    argsMap = {
        "logLevel": "WARN",
        "metricsToObtein": ["1", "2", "3", "4", "5", "6"]
    }

    for sParameter in parameters:
        if (sParameter.upper().find('LOGLEVEL=') != -1):
            argsMap["logLevel"] = sParameter.split("=")[1].upper()

        elif (sParameter.upper().find("METRICS=") != -1):
            argsMap["metricsToObtein"] = sParameter.split("=")[1].split(",")

    return argsMap

# Instantiate every metrics solver class and trigger the evaluations.
#
# @param moviesDataset  Dataset containing the movies information.
# @param ratingsDataset Dataset containing the ratings information.
# @param metrics        The metric ID to evaluate.
def runMetric(moviesDataset, ratingsDataset, metric):
    print("\n")

    if metric == '1':
        print("Best films by overall rating")
        return Metrics.BestFilmsByOverallRating(moviesDataset, ratingsDataset).collect()
    elif metric == '2':
        print("Most rated films")
        return Metrics.MostRatedFilms(moviesDataset, ratingsDataset).collect()
    elif metric == '3':
        print("Global ratings given by each user per category")
        return Metrics.GlobalRatingsGivenByEachUserPerCategory(moviesDataset, ratingsDataset).collect()
    elif metric == '4':
        print("Users giving the lowest ratings")
        return Metrics.UsersGivingTheLowestRatings(moviesDataset, ratingsDataset).collect()
    elif metric == '5':
        print("Genres by average rating")
        return Metrics.GenresByAverageRating(moviesDataset, ratingsDataset).collect()
    elif metric == '6':
        print("Genres combinations")
        return Metrics.GenresCombinations(moviesDataset, ratingsDataset).collect()
    else:
        raise Exception("Tried to obtain an unavailable metric: " + metric)

# Read file into dataframe using an existing schema.
#
# @param s        SparkSession.
# @param filename path to the file.
# @param schema   dataframe schema.
def readCsvIntoDataframe(s, filename, schema):
    csvPath = os.path.join(os.path.dirname(__file__), filename)
    df = s.read \
        .format("csv") \
        .option("header", "true") \
        .schema(schema) \
        .load(csvPath)

    return df

def printResults(results, dStart, dEnd):
    duration = dEnd - dStart
    print("Results obtained in {:.2f}s\n".format(duration))

    fieldNames = results[0].__fields__
    print("\t - \t".join((str(aFieldName) for aFieldName in fieldNames)))

    for row in results:
        print("\t - \t".join((str(r) for r in row)))


if __name__ == "__main__":
    main()
